### DISCUSSION TITLE
Multiple time range filters

### DISCUSSION THREAD

**Discussion by itziarmj** [apache/superset#24165] [Ideas]
In Superset 2.1.0 is possible to have multiple time-based filters but, as noted in https://preset.io/blog/exploring-trends-over-non-temporal-dimensions-in-superset-and-preset/#3-exploring-charts-in-dashboards, in the Dashboard there is only one time range for all of them.
Is it foreseen to have different time ranges for each of the time-based filters?
In our dashboards we have several time ranges to filter, like manufacturing date, delivery date... and it makes more sense to be able to filter them with different time ranges.
This was proposed also in https://github.com/apache/superset/issues/23842

**Comment by dijkstra2003** [comment]
Any updates on this? We would really like it and i'm considering building it myself

**Reply by sfirke** [reply]
I don't believe this has moved.  @dijkstra2003 would you be interested in contributing it as a feature back to the main Superset project?

**Reply by dijkstra2003** [reply]
If i'm building this i will definitely contribute and give the code to the main superset ;)

**Comment by goto-loop** [comment]
This would be a very useful feature. Another use case for it is a side-by-side comparison: having the same chart twice in one dashboard and applying a different time range to each.

**Reply by phyndman** [reply]
you can do this now by just applying each time filter to specific charts/panels using the "scoping" on the filter :-)

**Comment by Stefan781** [comment]
+1. Right now, if you add two time range filters, the second one seems to overwrites the first, which makes it impossible to recover its value even when using the _filter_ variable with Jinja.

**Comment by phyndman** [comment]
Don't suppose anyone found some clever workaround to inject 2 different "Time Ranges" using jinja into queries?

**Comment by axoxymous-coder** [comment]
Has anyone found / know any workaround on ways to have atleast two time-range filters in a dashboard? One of our use case is getting blocked since we need to filter on two date ranges ...

**Reply by Stefan781** [reply]
I think it is possible to have two time range filters but they must be assigned to two different charts from the Scoping menu.

**Reply by axoxymous-coder** [reply]
Hi @Stefan781 Thanks for your response. One of our use case requires two time range filters to be applied across all the charts in the dashboard. Seems this is not in Superset roadmap in 3.1.0 as per #27075

**Reply by Stefan781** [reply]
We are in the same boat. Our use case allowed for a workaround with a value filter (specify the number of months as a number, use Jinja to retrieve it and then transform it into a date) but if you want a true time range filter you need 6 such filters.

**Reply by phyndman** [reply]
I've created 4 "value" filters (Start A, End A, Start B, End B) that are actually text columns that look like dates, then inject these into metrics and calculated columns in the underlying datasets using Jinja which enable calculations like the below which you can show in the same chart etc.

Metrics...
- sum(case when order_date between "Start A" and "End A" then amount else 0 end)
- sum(case when order_date between "Start B" and "End B" then amount else 0 end)

Calc columns for barc charts/line splits etc...
- case when order_date between "Start A" and "End A" then 'A' end
- case when order_date between "Start B" and "End B" then 'B' end

**Reply by Stefan781** [reply]
@phyndman are you using predefined values for those filters?

**Reply by phyndman** [reply]
Yeh, essentially I've got a "date dimension" dataset and I've added 4 new columns to it (which are essentially replicas) that the filters get their values from so people can select them from a drop down. Not the most elegant thing but does what it says on the tin for the end users!

**Reply by Stefan781** [reply]
So are you limited to some predefined dates or did you come up with a workaround to let the user choose any date?

**Reply by phyndman** [reply]
User can choose any dates they want as essentially they get presented with a list of all dates in the 5 years prior to "today" in all 4 I've created and can select any date they want in each. They need to know that Start needs to be before End or else the query will be a bit strange however or may not work properly. With 2 proper time range filters they don't have to search for dates they could use the inbuilt functions on dates (last week, current month) so would just be a much easier user experience and better for them.

**Reply by Stefan781** [reply]
Aha, I see! That's a neat solution if the user doesn't mind scrolling to find their date. Thanks for sharing!
Edit: Just realized, that they don't have to scroll but they can also search for the value.

**Reply by phyndman** [reply]
Yeh lots of our use cases are looking at comparing any user defined "time range A" vs "time range B" in the same chart table and showing a variance. They can also apply more filters etc. Having to search though a massive list of dates is not ideal but gets to some sort of solution. If I were an awesome coder I'd have a go at being able to properly pull 2 time range filters into Jinja (currently you just get the one of them) but fear I'd bring down the entire project!!

**Comment by michael-s-molina** [comment]
Hi folks. I'm working on a SIP with @john-bodley and @kasiazjc to improve time filters that will resolve this problem. Just to set expectations, we are currently focused on delivering 4.0, but we should continue the SIP work after the release.

**Reply by phyndman** [reply]
Hi Michael. Just checking in on this to see if anything has been done on this?

**Reply by michael-s-molina** [reply]
Hi @phyndman. Unfortunately, priorities changed and we postponed this work for the 2025 roadmap.

**Reply by phyndman** [reply]
no worries thank you for the update. keep up the good work :-)

**Reply by Hatake07** [reply]
Hi @michael-s-molina , is there any public document on which we could take notes about roadmap milestones for 2025?

**Reply by michael-s-molina** [reply]
We generally use the [SIP board](https://github.com/orgs/apache/projects/170) to track the major epics that are happening. Currently, there's no one working on improving time filters.

**Reply by ichsan-data** [reply]
hi @michael-s-molina , is there anyupdate?

**Reply by michael-s-molina** [reply]
Hi @ichsan-data. This project was deprioritized in favor of Superset extensions. There's no current plan to tackle this in the short term.

**Reply by phanhuy1** [reply]
Hi @michael-s-molina  . Sorry but, any news on this matter? :(

**Comment by itziarmj** [comment]
We also have these types of cases. We have two workarounds:
When we need “flexible” dates, we let the user input the date as a string in a defined format. Pe 2024-01-01
When we have few options, we have a table with the list of possible dates to be selected.

**Comment by seandiacono** [comment]
+1, this would be super handy for us to be able to filter multiple time based attributes at the same time

**Comment by dtes** [comment]
I found that API can receive multiple time range filters. Similar to number range filters I send time ranges. Example of request I sent in my dashboard.
```
POST /api/v1/chart/data?form_data=%7B%22slice_id%22%3A95%7D&dashboard_id=114 HTTP/1.1
Host: superset.myapp.kz
Cookie: session=.eJyNVVuvozYQ_itVnruVbS4J-8YhgcABsuGOqyoCG8I1cBJIAqv-95qcbbVqX_qA
Content-Type: application/json
Authorization: Bearer eyJhbGciOiJIUzUxMiJ9.eyJzdWIiOiJhZG1pbiIsIm9yaWdprO98UA
Content-Length: 5949

{
    "datasource": {
      "id": 1,
      "type": "table"
    },
    "force": false,
    "queries": [
      {
        
        "filters": [
          {
            "col": "doc_category_id",
            "op": ">=",
            "val": 0
          },
          {
            "col": "doccreatedata_dtcreate",
            "op": ">=",
            "val": "2024-12-08T00:00:00"
          },
          {
            "col": "doccreatedata_dtcreate",
            "op": "<=",
            "val": "2024-12-10T00:00:00"
          },
          {
            "col": "regdata_regdt",
            "op": ">=",
            "val": "2024-12-10T00:00:00"
          },
          {
            "col": "regdata_regdt",
            "op": "<=",
            "val": "2024-12-13T00:00:00"
          },
          {
            "col": "doccreatedata_dtcreate",
            "op": "TEMPORAL_RANGE",
            "val": "No filter"
          },
          {
            "col": "doc_type_name",
            "op": "IN",
            "val": [
              "IncomeDoc"
            ]
          },
          {
            "col": "regdata_regdt",
            "op": "TEMPORAL_RANGE",
            "val": "No filter"
          }
        ],
        "extras": {
          "time_grain_sqla": "P1D",
          "having": "",
          "where": ""
        },
        "applied_time_extras": {},
        "columns": [
          {
            "expressionType": "SQL",
            "label": "Ссылка",
            "sqlExpression": "CONCAT('<a href=\"', 'https://myapp.kz/app/view/', id,'\" target=\"_blank\">Посмотреть документ</a>')"
          },
          {
            "expressionType": "SQL",
            "label": "Номер документа",
            "sqlExpression": "doc_num_calc || ''"
          },
          {
            "expressionType": "SQL",
            "label": "Тип документа",
            "sqlExpression": "controldata_controltype_labelru || ''"
          },
          "doc_type_labelru",
          "regdata_regdt",
          {
            "expressionType": "SQL",
            "label": "Рег дата или дата создания",
            "sqlExpression": "case \nwhen regdata_regdt is not null\nthen regdata_regdt\nwhen regdata_regdt is  null\nthen doccreatedata_dtcreate\nend"
          }
        ],
        "orderby": [],
        "annotation_layers": [],
        "row_limit": 1000,
        "series_limit": 0,
        "order_desc": true,
        "url_params": {},
        "custom_params": {},
        "custom_form_data": {},
        "post_processing": []
      }
    ],
    "form_data": {
      "datasource": "1__table",
      "viz_type": "table",
      "slice_id": 95,
      "query_mode": "raw",
      "groupby": [],
      "time_grain_sqla": "P1D",
      "temporal_columns_lookup": {
        "regdata_regdt": true,
        "controldata_initialduedt": true,
        "controldata_duedt": true,
        "controldata_controlenddt": true,
        "paymentdata_paymentregisterdt": true,
        "doccreatedata_dtcreate": true
      },
      "all_columns": [
        {
          "expressionType": "SQL",
          "label": "Ссылка",
          "sqlExpression": "CONCAT('<a href=\"', 'https://myapp.kz/app/view/', id,'\" target=\"_blank\">Посмотреть документ</a>')"
        },
        {
          "expressionType": "SQL",
          "label": "Номер документа",
          "sqlExpression": "doc_num_calc || ''"
        },
        {
          "expressionType": "SQL",
          "label": "Тип документа",
          "sqlExpression": "controldata_controltype_labelru || ''"
        },
        "doc_type_labelru",
        "regdata_regdt",
        {
          "expressionType": "SQL",
          "label": "Рег дата или дата создания",
          "sqlExpression": "case \nwhen regdata_regdt is not null\nthen regdata_regdt\nwhen regdata_regdt is  null\nthen doccreatedata_dtcreate\nend"
        }
      ],
      "percent_metrics": [],
      "adhoc_filters": [
        {
          "expressionType": "SIMPLE",
          "subject": "doccreatedata_dtcreate",
          "operator": "TEMPORAL_RANGE",
          "comparator": "No filter",
          "clause": "WHERE",
          "sqlExpression": null,
          "isExtra": false,
          "isNew": false,
          "datasourceWarning": false,
          "filterOptionName": "filter_ch5hs924zt6_qrxq3tj3asj"
        },
        {
          "expressionType": "SIMPLE",
          "subject": "doc_type_name",
          "operator": "IN",
          "operatorId": "IN",
          "comparator": [
            "IncomeDoc"
          ],
          "clause": "WHERE",
          "sqlExpression": null,
          "isExtra": false,
          "isNew": false,
          "datasourceWarning": false,
          "filterOptionName": "filter_v3k42b054eh_ezigirufv3w"
        },
        {
          "expressionType": "SIMPLE",
          "subject": "regdata_regdt",
          "operator": "TEMPORAL_RANGE",
          "operatorId": "TEMPORAL_RANGE",
          "comparator": "No filter",
          "clause": "WHERE",
          "sqlExpression": null,
          "isExtra": false,
          "isNew": false,
          "datasourceWarning": false,
          "filterOptionName": "filter_wbpeaftqyab_b8l5kxul8bv"
        }
      ],
      "order_by_cols": [],
      "row_limit": 1000,
      "server_page_length": 10,
      "order_desc": true,
      "table_timestamp_format": "smart_date",
      "show_cell_bars": true,
      "color_pn": true,
      "extra_form_data": {
        "filters": [
          {
            "col": "doc_category_id",
            "op": ">=",
            "val": 29
          }
        ],
        "time_range": "2024-12-08T00:00:00 : 2024-12-12T00:00:00"
      },
      "dashboards": [
        77
      ],
      "label_colors": {},
      "shared_label_colors": {},
      "extra_filters": [],
      "dashboardId": 114,
      "force": false,
      "result_format": "json",
      "result_type": "full",
      "include_time": false
    },
    "result_format": "json",
    "result_type": "full"
  }
```

See how I send time range filters:
```
          {
            "col": "doccreatedata_dtcreate",
            "op": ">=",
            "val": "2024-12-08T00:00:00"
          },
          {
            "col": "doccreatedata_dtcreate",
            "op": "<=",
            "val": "2024-12-10T00:00:00"
          },
          {
            "col": "regdata_regdt",
            "op": ">=",
            "val": "2024-12-10T00:00:00"
          },
          {
            "col": "regdata_regdt",
            "op": "<=",
            "val": "2024-12-13T00:00:00"
          }
```

In response, you can get a generated SQL query, and in my case, this is the correct query.

So, we need to add custom time range filters to the superset frontend and chose it in the dashboard filters.

**Comment by marianysilva** [comment]
Use case:

- Date Range: Last Week

- Chart A:
    - Dataset: 1
    - Column X in Date Range filter
- Chart B:
    - Dataset: 1
    - Column Y in Date Range filter
  
- Details Table:
    - Dataset: 1
    - Column X OR Column Y in Date Range filter
  
<img width="876" height="792" alt="Screenshot 2025-09-16 at 11 55 10" src="https://github.com/user-attachments/assets/2ce80bfa-1ae5-4994-aacf-97a66d9f24a5" />

**Comment by ThanhNg** [comment]
Upvote
